[bug] AssertionError: Conan failed to acquire database lock

### Describe the bug

Windows Server 2022, Conan 2.22.1, Python 3.11

The problem arises after upgrade Conan from 2.21.0 to 2.22.0 (and 2.22.1) on our CI workflow during `conan install` packages. It's not quite stable fail, sometimes it works as expected. Below you can find logs (some parts are truncated for brevity):

```
Run py -3 -m pip install conan==2.22.1
Collecting conan==2.22.1
  Downloading conan-2.22.1.tar.gz (542 kB)
     -------------------------------------- 542.4/542.4 kB 3.8 MB/s eta 0:00:00
  Installing build dependencies: started
  Installing build dependencies: finished with status 'done'
  Getting requirements to build wheel: started
  Getting requirements to build wheel: finished with status 'done'
  Preparing metadata (pyproject.toml): started
  Preparing metadata (pyproject.toml): finished with status 'done'
Requirement already satisfied: requests<3.0.0,>=2.25 in c:\python311\lib\site-packages (from conan==2.22.1) (2.32.4)
Requirement already satisfied: urllib3<3.0,>=1.26.6 in c:\python311\lib\site-packages (from conan==2.22.1) (1.26.20)
Requirement already satisfied: colorama<0.5.0,>=0.4.3 in c:\python311\lib\site-packages (from conan==2.22.1) (0.4.6)
Requirement already satisfied: PyYAML<7.0,>=6.0 in c:\python311\lib\site-packages (from conan==2.22.1) (6.0.2)
Collecting patch-ng<1.19,>=1.18.0 (from conan==2.22.1)
  Downloading patch-ng-1.18.1.tar.gz (17 kB)
  Installing build dependencies: started
  Installing build dependencies: finished with status 'done'
  Getting requirements to build wheel: started
  Getting requirements to build wheel: finished with status 'done'
  Preparing metadata (pyproject.toml): started
  Preparing metadata (pyproject.toml): finished with status 'done'
Requirement already satisfied: fasteners>=0.15 in c:\python311\lib\site-packages (from conan==2.22.1) (0.19)
Requirement already satisfied: Jinja2<4.0.0,>=3.0 in c:\python311\lib\site-packages (from conan==2.22.1) (3.1.6)
Requirement already satisfied: python-dateutil<3,>=2.8.0 in c:\python311\lib\site-packages (from conan==2.22.1) (2.9.0.post0)
Requirement already satisfied: MarkupSafe>=2.0 in c:\python311\lib\site-packages (from Jinja2<4.0.0,>=3.0->conan==2.22.1) (3.0.2)
Requirement already satisfied: six>=1.5 in c:\python311\lib\site-packages (from python-dateutil<3,>=2.8.0->conan==2.22.1) (1.16.0)
Requirement already satisfied: charset_normalizer<4,>=2 in c:\python311\lib\site-packages (from requests<3.0.0,>=2.25->conan==2.22.1) (3.4.2)
Requirement already satisfied: idna<4,>=2.5 in c:\python311\lib\site-packages (from requests<3.0.0,>=2.25->conan==2.22.1) (3.10)
Requirement already satisfied: certifi>=2017.4.17 in c:\python311\lib\site-packages (from requests<3.0.0,>=2.25->conan==2.22.1) (2025.7.14)
Building wheels for collected packages: conan, patch-ng
  Building wheel for conan (pyproject.toml): started
  Building wheel for conan (pyproject.toml): finished with status 'done'
  Created wheel for conan: filename=conan-2.22.1-py3-none-any.whl size=682077 sha256=4390d9d82d06611c2acd6e87abcd7927a05efd5e4e740782616012df23e02704
  Stored in directory: c:\users\gh-ci\appdata\local\pip\cache\wheels\59\1b\43\6502c708257c50b84469763d7f15a80db57489b28b5c5c3db7
  Building wheel for patch-ng (pyproject.toml): started
  Building wheel for patch-ng (pyproject.toml): finished with status 'done'
  Created wheel for patch-ng: filename=patch_ng-1.18.1-py3-none-any.whl size=17028 sha256=cdc3c7269eeb25b966e061f19daf34445608598f14f8bc54b57b3127a51b1f1a
  Stored in directory: c:\users\gh-ci\appdata\local\pip\cache\wheels\7b\cc\49\8bc9f301f7969862da20988c1bd25c60f20df61718a01b2449
Successfully built conan patch-ng
Installing collected packages: patch-ng, conan
  Attempting uninstall: patch-ng
    Found existing installation: patch-ng 1.17.4
    Uninstalling patch-ng-1.17.4:
      Successfully uninstalled patch-ng-1.17.4
  Attempting uninstall: conan
    Found existing installation: conan 1.61.0
    Uninstalling conan-1.61.0:
      Successfully uninstalled conan-1.61.0
Successfully installed conan-2.22.1 patch-ng-1.18.1

Notice:  A new release of pip is available: 23.2.1 -> 25.3
Notice:  To update, run: python.exe -m pip install --upgrade pip

....

======== Installing packages ========

-------- Downloading 95 packages --------
Downloading binary packages in 16 parallel threads


....



ERROR: Traceback (most recent call last):
  File "C:\Python311\Lib\site-packages\conan\cli\cli.py", line 297, in main
    cli.run(args)
  File "C:\Python311\Lib\site-packages\conan\cli\cli.py", line 193, in run
    command.run(self._conan_api, args[0][1:])
  File "C:\Python311\Lib\site-packages\conan\cli\command.py", line 191, in run
    info = self._method(conan_api, parser, *args)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "C:\Python311\Lib\site-packages\conan\cli\commands\install.py", line 75, in install
    conan_api.install.install_binaries(deps_graph=deps_graph, remotes=remotes)
  File "C:\Python311\Lib\site-packages\conan\api\subapi\install.py", line 30, in install_binaries
    installer.install(deps_graph, remotes, install_order=install_order)
  File "C:\Python311\Lib\site-packages\conan\internal\graph\installer.py", line 255, in install
    self._download_bulk(install_order)
  File "C:\Python311\Lib\site-packages\conan\internal\graph\installer.py", line 286, in _download_bulk
    thread_pool.map(self._download_pkg, downloads)
  File "C:\Python311\Lib\multiprocessing\pool.py", line 367, in map
    return self._map_async(func, iterable, mapstar, chunksize).get()
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "C:\Python311\Lib\multiprocessing\pool.py", line 774, in get
    raise self._value
  File "C:\Python311\Lib\multiprocessing\pool.py", line 125, in worker
    result = (True, func(*args, **kwds))
                    ^^^^^^^^^^^^^^^^^^^
  File "C:\Python311\Lib\multiprocessing\pool.py", line 48, in mapstar
    return list(map(*args))
           ^^^^^^^^^^^^^^^^
  File "C:\Python311\Lib\site-packages\conan\internal\graph\installer.py", line 297, in _download_pkg
    self._remote_manager.get_package(node.pref, node.binary_remote)
  File "C:\Python311\Lib\site-packages\conan\internal\rest\remote_manager.py", line 142, in get_package
    pkg_layout = self._cache.create_pkg_layout(pref)
                 ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "C:\Python311\Lib\site-packages\conan\internal\cache\cache.py", line 161, in create_pkg_layout
    self._db.create_package(package_path, pref, None)
  File "C:\Python311\Lib\site-packages\conan\internal\cache\db\cache_database.py", line 87, in create_package
    self._packages.create(path, ref, build_id=build_id)
  File "C:\Python311\Lib\site-packages\conan\internal\cache\db\packages_table.py", line 80, in create
    with self.db_connection() as conn:
  File "C:\Python311\Lib\contextlib.py", line 137, in __enter__
    return next(self.gen)
           ^^^^^^^^^^^^^^
  File "C:\Python311\Lib\site-packages\conan\internal\cache\db\table.py", line 26, in db_connection
    assert self._lock.acquire(timeout=20), "Conan failed to acquire database lock"
AssertionError: Conan failed to acquire database lock

ERROR: Conan failed to acquire database lock
```

It seems highly likely related to parallel download, that is used by default since 2.22 (https://github.com/conan-io/conan/pull/19099). I noticed similar fails before on 2.21 when I was experimenting with `--core-conf core.download:parallel={os.cpu_count()}` for `conan install` command.

If it's necessary to provide additional details to simplify investigation, please let me know.

I found old similar bug that is marked as fixed: https://github.com/conan-io/conan/issues/13924

BTW: Will downloading work in single-thread mode if I add `--core-conf core.download:parallel=1`? Instead if I'm not able to disable parallel download we will stuck on 2.21 version, due to that error with database lock.

--- a/conan/api/subapi/download.py
+++ b/conan/api/subapi/download.py
@@ -24,13 +24,13 @@ class DownloadAPI:
         app = ConanBasicApp(self._conan_api)
         assert ref.revision, f"Reference '{ref}' must have revision"
         try:
-            app.cache.recipe_layout(ref)  # raises if not found
+            recipe_layout = app.cache.recipe_layout(ref)  # raises if not found
         except ConanException:
             pass
         else:
             output.info(f"Skip recipe {ref.repr_notime()} download, already in cache")
             if metadata:
-                app.remote_manager.get_recipe_metadata(ref, remote, metadata)
+                app.remote_manager.get_recipe_metadata(recipe_layout, ref, remote, metadata)
             return False
 
         output.info(f"Downloading recipe '{ref.repr_notime()}'")
@@ -40,11 +40,10 @@ class DownloadAPI:
             server_ref = app.remote_manager.get_recipe_revision(ref, remote)
             assert server_ref == ref
             ref.timestamp = server_ref.timestamp
-        app.remote_manager.get_recipe(ref, remote, metadata)
+        recipe_layout = app.remote_manager.get_recipe(ref, remote, metadata)
 
         # Download the sources too, don't be lazy
         output.info(f"Downloading '{str(ref)}' sources")
-        recipe_layout = app.cache.recipe_layout(ref)
         app.remote_manager.get_recipe_sources(ref, recipe_layout, remote)
         return True
 

--- a/conan/api/subapi/upload.py
+++ b/conan/api/subapi/upload.py
@@ -7,8 +7,7 @@ from conan.api.model import PackagesList, Remote
 from conan.api.output import ConanOutput
 from conan.internal.api.upload import add_urls
 from conan.internal.conan_app import ConanApp
-from conan.internal.api.uploader import PackagePreparator, UploadExecutor, UploadUpstreamChecker, \
-    gather_metadata
+from conan.internal.api.uploader import PackagePreparator, UploadExecutor, UploadUpstreamChecker
 from conan.internal.rest.pkg_sign import PkgSignaturesPlugin
 from conan.internal.rest.file_uploader import FileUploader
 from conan.internal.errors import AuthenticationException, ForbiddenException
@@ -66,9 +65,7 @@ class UploadAPI:
             raise ConanException("Empty string and patterns can not be mixed for metadata.")
         app = ConanApp(self._conan_api)
         preparator = PackagePreparator(app, self._api_helpers.global_conf)
-        preparator.prepare(package_list, enabled_remotes)
-        if metadata != ['']:
-            gather_metadata(package_list, app.cache, metadata)
+        preparator.prepare(package_list, enabled_remotes, metadata)
         signer = PkgSignaturesPlugin(app.cache, app.cache_folder)
         # This might add files entries to package_list with signatures
         signer.sign(package_list)

--- a/conan/internal/api/uploader.py
+++ b/conan/internal/api/uploader.py
@@ -113,11 +113,11 @@ class PackagePreparator:
         self._compressformat = compressformat
         self._compresslevel = compresslevel
 
-    def prepare(self, pkg_list, enabled_remotes):
+    def prepare(self, pkg_list, enabled_remotes, metadata):
         local_url = self._global_conf.get("core.scm:local_url", choices=["allow", "block"])
         for ref, packages in pkg_list.items():
-            layout = self._app.cache.recipe_layout(ref)
-            conanfile_path = layout.conanfile()
+            recipe_layout = self._app.cache.recipe_layout(ref)
+            conanfile_path = recipe_layout.conanfile()
             conanfile = self._app.loader.load_basic(conanfile_path)
             url = conanfile.conan_data.get("scm", {}).get("url") if conanfile.conan_data else None
             if local_url != "allow" and url is not None:
@@ -131,21 +131,30 @@ class PackagePreparator:
             bundle.pop("files", None)
             bundle.pop("upload-urls", None)
             if bundle.get("upload"):
-                self._prepare_recipe(ref, bundle, conanfile, enabled_remotes)
+                self._prepare_recipe(recipe_layout, ref, bundle, conanfile, enabled_remotes)
+
+            # Package metadata files too
+            if metadata != [""] and (metadata or bundle.get("upload")):
+                metadata_folder = recipe_layout.metadata()
+                files = _metadata_files(metadata_folder, metadata)
+                if files:
+                    ConanOutput(scope=str(ref)).info(f"Recipe metadata: {len(files)} files")
+                    bundle.setdefault("files", {}).update(files)
+                    bundle["upload"] = True
+
             for pref in packages:
                 prev_bundle = pkg_list.package_dict(pref)
                 prev_bundle.pop("files", None)  # If defined from a previous upload
                 prev_bundle.pop("upload-urls", None)
                 if prev_bundle.get("upload"):
-                    self._prepare_package(pref, prev_bundle)
+                    self._prepare_package(pref, prev_bundle, metadata)
 
-    def _prepare_recipe(self, ref, ref_bundle, conanfile, remotes):
+    def _prepare_recipe(self, recipe_layout, ref, ref_bundle, conanfile, remotes):
         """ do a bunch of things that are necessary before actually executing the upload:
         - retrieve exports_sources to complete the recipe if necessary
         - compress the artifacts in conan_export.tgz and conan_export_sources.tgz
         """
         try:
-            recipe_layout = self._app.cache.recipe_layout(ref)
             retrieve_exports_sources(self._app.remote_manager, recipe_layout, conanfile, ref,
                                      remotes)
             cache_files = self._compress_recipe_files(recipe_layout, ref)
@@ -188,13 +197,22 @@ class PackagePreparator:
             result[comp] = os.path.join(download_export_folder, comp)
         return result
 
-    def _prepare_package(self, pref, prev_bundle):
+    def _prepare_package(self, pref, prev_bundle, metadata):
         pkg_layout = self._app.cache.pkg_layout(pref)
         if pkg_layout.package_is_dirty():
             raise ConanException(f"Package {pref} is corrupted, aborting upload.\n"
                                  f"Remove it with 'conan remove {pref}'")
         cache_files = self._compress_package_files(pkg_layout, pref)
         prev_bundle["files"] = cache_files
+
+        # Package metadata files too
+        if metadata != [""] and (metadata or prev_bundle["upload"]):
+            metadata_folder = pkg_layout.metadata()
+            files = _metadata_files(metadata_folder, metadata)
+            if files:
+                ConanOutput(scope=str(pref)).info(f"Package metadata: {len(files)} files")
+                prev_bundle.setdefault("files", {}).update(files)
+                prev_bundle["upload"] = True
 
     def _compressed_file(self, filename, files, download_folder, ref):
         output = ConanOutput(scope=str(ref))
@@ -368,25 +386,3 @@ def _metadata_files(folder, metadata):
             path = os.path.join("metadata", relpath).replace("\\", "/")
             result[path] = abs_path
     return result
-
-
-def gather_metadata(package_list, cache, metadata):
-    for rref, packages in package_list.items():
-        recipe_bundle = package_list.recipe_dict(rref)
-        if metadata or recipe_bundle["upload"]:
-            metadata_folder = cache.recipe_layout(rref).metadata()
-            files = _metadata_files(metadata_folder, metadata)
-            if files:
-                ConanOutput(scope=str(rref)).info(f"Recipe metadata: {len(files)} files")
-                recipe_bundle.setdefault("files", {}).update(files)
-                recipe_bundle["upload"] = True
-
-        for pref in packages:
-            pkg_bundle = package_list.package_dict(pref)
-            if metadata or pkg_bundle["upload"]:
-                metadata_folder = cache.pkg_layout(pref).metadata()
-                files = _metadata_files(metadata_folder, metadata)
-                if files:
-                    ConanOutput(scope=str(pref)).info(f"Package metadata: {len(files)} files")
-                    pkg_bundle.setdefault("files", {}).update(files)
-                    pkg_bundle["upload"] = True

--- a/conan/internal/api/uploader.py
+++ b/conan/internal/api/uploader.py
@@ -146,8 +146,7 @@ class PackagePreparator:
                 prev_bundle = pkg_list.package_dict(pref)
                 prev_bundle.pop("files", None)  # If defined from a previous upload
                 prev_bundle.pop("upload-urls", None)
-                if prev_bundle.get("upload"):
-                    self._prepare_package(pref, prev_bundle, metadata)
+                self._prepare_package(pref, prev_bundle, metadata)
 
     def _prepare_recipe(self, recipe_layout, ref, ref_bundle, conanfile, remotes):
         """ do a bunch of things that are necessary before actually executing the upload:
@@ -198,15 +197,18 @@ class PackagePreparator:
         return result
 
     def _prepare_package(self, pref, prev_bundle, metadata):
-        pkg_layout = self._app.cache.pkg_layout(pref)
-        if pkg_layout.package_is_dirty():
-            raise ConanException(f"Package {pref} is corrupted, aborting upload.\n"
-                                 f"Remove it with 'conan remove {pref}'")
-        cache_files = self._compress_package_files(pkg_layout, pref)
-        prev_bundle["files"] = cache_files
+        pkg_layout = None
+        if prev_bundle.get("upload"):
+            pkg_layout = self._app.cache.pkg_layout(pref)
+            if pkg_layout.package_is_dirty():
+                raise ConanException(f"Package {pref} is corrupted, aborting upload.\n"
+                                     f"Remove it with 'conan remove {pref}'")
+            cache_files = self._compress_package_files(pkg_layout, pref)
+            prev_bundle["files"] = cache_files
 
         # Package metadata files too
-        if metadata != [""] and (metadata or prev_bundle["upload"]):
+        if metadata != [""] and (metadata or prev_bundle.get("upload")):
+            pkg_layout = pkg_layout or self._app.cache.pkg_layout(pref)
             metadata_folder = pkg_layout.metadata()
             files = _metadata_files(metadata_folder, metadata)
             if files:

--- a/conan/internal/graph/installer.py
+++ b/conan/internal/graph/installer.py
@@ -115,12 +115,9 @@ class _PackageBuilder:
         # FIXME: Conan 2.0 Clear the registry entry (package ref)
         return prev
 
-    def build_package(self, node, package_layout):
+    def build_package(self, node, recipe_layout, package_layout):
         conanfile = node.conanfile
         pref = node.pref
-
-        # TODO: cache2.0 fix this
-        recipe_layout = self._cache.recipe_layout(pref.ref)
 
         base_source = recipe_layout.source()
         base_package = package_layout.package()
@@ -198,6 +195,7 @@ class BinaryInstaller:
         conanfile.folders.set_base_export_sources(source_folder)
         conanfile.folders.set_base_recipe_metadata(recipe_layout.metadata())
         config_source(export_source_folder, conanfile, self._hook_manager)
+        return recipe_layout
 
     @staticmethod
     def install_system_requires(graph, only_info=False, install_order=None):
@@ -257,8 +255,9 @@ class BinaryInstaller:
         for level in install_order:
             for install_reference in level:
                 for package in install_reference.packages.values():
-                    self._install_source(package.nodes[0], remotes)
-                    self._handle_package(package, install_reference, handled_count, package_count)
+                    recipe_layout = self._install_source(package.nodes[0], remotes)
+                    self._handle_package(recipe_layout, package, install_reference, handled_count,
+                                         package_count)
                     handled_count += 1
                     if package.binary == BINARY_CACHE:
                         prefs_lru.append(package.nodes[0].pref)
@@ -301,7 +300,7 @@ class BinaryInstaller:
         assert node.pref.timestamp is not None
         self._remote_manager.get_package(node.pref, node.binary_remote)
 
-    def _handle_package(self, package, install_reference, handled_count, total_count):
+    def _handle_package(self, recipe_layout, package, install_reference, handled_count, total_count):
         if package.binary in (BINARY_EDITABLE, BINARY_EDITABLE_BUILD):
             self._handle_node_editable(package)
             return
@@ -322,7 +321,7 @@ class BinaryInstaller:
             for line in compact_dumps:
                 ConanOutput(scope=str(pref.ref)).info(line, fg=Color.BRIGHT_GREEN)
             package_layout = self._cache.create_build_pkg_layout(pref)
-            self._handle_node_build(package, package_layout)
+            self._handle_node_build(package, recipe_layout, package_layout)
             # Just in case it was recomputed
             package.package_id = package.nodes[0].pref.package_id  # Just in case it was recomputed
             package.prev = package.nodes[0].pref.revision
@@ -388,7 +387,7 @@ class BinaryInstaller:
             # TODO: Check this base_path usage for editable when not defined
             self._call_package_info(conanfile, package_folder=rooted_base_path, is_editable=True)
 
-    def _handle_node_build(self, package, pkg_layout):
+    def _handle_node_build(self, package, recipe_layout, pkg_layout):
         node = package.nodes[0]
         pref = node.pref
         assert pref.package_id, "Package-ID without value"
@@ -400,7 +399,7 @@ class BinaryInstaller:
             pkg_layout.package_remove()
             with pkg_layout.set_dirty_context_manager():
                 builder = _PackageBuilder(self._app, self._hook_manager)
-                pref = builder.build_package(node, pkg_layout)
+                pref = builder.build_package(node, recipe_layout, pkg_layout)
             assert node.prev, "Node PREV shouldn't be empty"
             assert node.pref.revision, "Node PREF revision shouldn't be empty"
             assert pref.revision is not None, "PREV for %s to be built is None" % str(pref)

--- a/conan/internal/rest/remote_manager.py
+++ b/conan/internal/rest/remote_manager.py
@@ -102,15 +102,14 @@ class RemoteManager:
         mkdir(layout.metadata())
         return layout
 
-    def get_recipe_metadata(self, ref, remote, metadata):
+    def get_recipe_metadata(self, recipe_layout, ref, remote, metadata):
         """
         Get only the metadata for a locally existing recipe in Cache
         """
         assert ref.revision, "get_recipe without revision specified"
         output = ConanOutput(scope=str(ref))
         output.info("Retrieving recipe metadata from remote '%s' " % remote.name)
-        layout = self._cache.recipe_layout(ref)
-        download_export = layout.download_export()
+        download_export = recipe_layout.download_export()
         try:
             self._call_remote(remote, "get_recipe", ref, download_export, metadata,
                               only_metadata=True)

